Gestion des exports d'un contenu

AmauryCarrade opened the pull request:
Ce commit ajoute une interface pour gérer les exports d'un contenu, en connaître l'état de traitement, et relancer, en cas de nécessité, leur traitement.

Il ajoute également une API pour lister les exports d'un contenu et leurs URLs.

Numéro du ticket concerné (optionnel) : closes #5830.

Le ticket d'origine parlait d'une page listant _toutes les demandes d'export passées_, mais après discussion avec @Situphen son auteur, nous sommes parvenus à la conclusion que ce n'était pas si pertinent que cela. À la place, cette proposition ajoute une interface listant uniquement la dernière demande d'export pour chaque type, ainsi que son statut (en attente, en cours de traitement, traitée, ou échouée), ainsi qu'un moyen de relancer la génération si nécessaire (ce que faisait initialement le lien dans la barre latérale).

L'interface étant dynamique (dans une modale, sans rechargement de la page), cette PR ajoute également une API pour récupérer les derniers exports d'un contenu. Cette API est _publique_, car j'estime qu'il n'est pas nécessaire de la protéger, mais ça pourrait se discuter.

![L'interface de gestion des exports](https://user-images.githubusercontent.com/1417570/85011145-6e60c080-b161-11ea-872f-3d31532add21.png)

L'interface se met à jour en temps réel lorsque nécessaire (i.e. lorsqu'elle est ouverte _et_ qu'il y a au moins un export en cours).

Des tests ont été ajoutés — bien entendu —, mais également pour l'API pré-existante de demande d'exports, car elle n'était pas testée jusqu'à présent.

#### Cette PR ajoute une requête SQL brute

L'API listant les derniers exports par type utilise une requête SQL brute (`PublicationEvent.objects.raw('…')`), car Django ne supporte pas la sélection depuis une sous-requête (`SELECT * FROM (SELECT …) WHERE …`).

### Contrôle qualité

Il y a deux points à tester : le fonctionnement correct de l'interface, et le fonctionnement de l'optimisation des requêtes de mise à jour.

Comme pré-requis à tous les tests, vous devez avoir une installation de ZdS et deux terminaux : un pour lancer le site et un pour le _watchdog_.

#### Notes préliminaires

Si vous n'avez jamais lancé le watchdog sur votre instance et qu'elle est relativement récente, il est possible que vous ayez une grande quantité de contenus en attente de traitement par ce dernier. Afin d'éviter d'attendre une heure qu'ils soient tous traités, vous pouvez éliminer toutes ces demandes avant de commencer la QA. Ouvrez un _shell_ Django avec la commande `python manage.py shell` depuis l'environnement virtuel, puis : 

```py
Python 3.8.3 (default, May 17 2020, 18:15:42) 
[GCC 10.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> from zds.tutorialv2.models.database import PublicationEvent
>>> PublicationEvent.objects.all().delete()
(0, {'tutorialv2.PublicationEvent': 0})
```

Cette suppression n'a aucune incidence nocive sur votre instance ni sur ses exports : elle supprime uniquement l'historique des _demandes d'export_, mais les éventuels fichiers restent toujours là.

Autre point : il est probable que sur votre instance locale, l'export PDF échoue. Cela se produit car vous n'avez pas tous les fichiers LaTeX (et notamment le _template_) pour générer ce dernier. Ce n'est absolument pas grave, car nous ne testons pas ici le fonctionnement du système d'export, mais uniquement son interface. C'est même bien, car cela permet de s'assurer que ce dernier réagit bien en cas d'erreur.

#### Test de l'interface

- Lancez le site avec `make run`, ce qui aura pour effet de reconstruire le _front_ au passage. Pour l'instant, ne lancez pas le _watchdog_.
- Créez et publiez un contenu. Vous pouvez également utiliser un contenu existant généré par les _fixtures_, peu importe.
- Sur la version brouillon du contenu publié, **cliquez sur le lien « Gérer les exports »** dans la barre latérale. Une interface modale doit s'ouvrir tel que montré plus haut.
   - _**Note** – Selon si vous avez ou non supprimé les demandes d'export et si vous utilisez un contenu fraîchement publié ou anciennement présent, et supposant que vous n'avez jamais lancé le _watchdog_ avant, vous verrez ou bien tous les exports comme étant « En attente », ou bien aucun export du tout.)_
- **Cliquez sur le bouton vert « Demander la génération de tous les exports ».** Après un temps de chargement, tous les exports doivent passer « En attente », et la date indiquée est actualisée (pour vérifier que la minute est bonne, la date complète apparaît au survol).
- Maintenant, **lancez le _watchdog_**, sans quitter la page ni la modale, ni l'actualiser ou quoique ce soit. Vous devriez observer que l'interface se met à jour en temps réel au fur et à mesure du traitement par le _watchdog_ : les exports passent à l'état _en traitement_, puis _terminé_ ou _échoué_, le cas échéant.
   - _**Note** – Il est très possible que l'interface n'affiche occasionnellement plus les exports quand le watchdog est en train de traiter des fichiers. Ce n'est **pas** un problème : ça vient du fait qu'une base SQLite supporte moins bien les accès concurrents que le SGBD que l'on utilise en production (si vous regardez les erreurs remontées, il s'agit de « database is locked »). Il est également possible que le watchdog plante à cause de cela : dans ce cas, relancez-le._
- Si un export échoue, le lien de téléchargement se transforme en un lien de relancement. Cliquer dessus relance tous les exports.
   - _**Note** – Ce test peut être combiné avec la section suivante, qui impliquera également de relancer un export à un moment, afin de gagner du temps._
- Maintenant, allez sur la version publique du contenu. Le lien de gestion des exports doit être présent et l'interface doit s'afficher à l'identique. (Inutile de re-tester le reste.)

#### Tests de l'optimisation de la mise à jour en temps réel

L'interface se met à jour à l'aide d'un système de requêtes régulières à l'API. Ce système ne tourne que lorsque strictement nécessaire, pour éviter toute requête superflue.

Pour cette partie de la QA, je vous invite à couper le _watchdog_ : ce sera plus simple.

- Rendez-vous sur la page d'un contenu pour lequel vous avez accès à la gestion des exports. Assurez-vous que les exports sont tous traités (ce qui devrait être le cas si vous faites cette QA dans l'ordre).
- **Ouvrez les outils de développement de votre navigateur**, à la section « Réseau » (`F12` ou `Ctrl + Maj + I`, généralement, ou `Ctrl + Maj + E` sous Firefox pour accéder directement à l'onglet réseau).
- **Ouvrez l'interface de gestion des exports.** Dans la console réseau, vous voyez une seule requête.
- **Cliquez sur le bouton de demande de tous les exports.** En plus de la requête `POST` d'envoi de la demande, vous constatez qu'il y a une requête `GET` envoyée toutes les quelques secondes.
- **Fermez l'interface de gestion des exports** (sans quitter la page pour autant). Vous observez que même si le traitement est toujours en cours, les requêtes cessent.
- Maintenant, **relancez le _watchdog_ et rouvrez l'interface**. Vous constatez que les requêtes régulières reprennent (même si certaines échouent peut-être).
- Dés que le traitement est achevé (tous les exports sont à « Terminé » ou « Échoué »), les requêtes cessent.

AmauryCarrade committed:
Gestion des exports d'un contenu

Ce commit ajoute une interface pour gérer les exports d'un contenu, en connaître
l'état de traitement, et relancer, en cas de nécessité, leur traitement.

Il ajoute également une API pour lister les exports d'un contenu et leurs URLs.

AmauryCarrade committed:
Optimisation des requêtes & test du nombre de requêtes effectuées.

- L'API listant les exports étant utilisée pour faire de la mise à jour en temps
  réel, on vérifie qu'il n'y a bien une seule requête effectuée.
- L'ajout de ce test a permis de remarquer que cinq requêtes étaient effectuées
  au lieu de deux utiles (quatre étant identiques entre elles) ; un prefetch a
  été ajouté pour éviter cela.

Situphen commented on assets/scss/pages/_content-exports.scss:
```diff
@@ -0,0 +1,136 @@
+                    color: #6c6c6c;
+
+                    a {
+                        //color: $color-primary;
```
Commentaire à enlever ?

Situphen commented on zds/tutorialv2/api/tests.py:
```diff
@@ -188,3 +189,101 @@ def test_get_content_reaction_voters(self):
+
+    def tearDown(self):
+        pass
+        # if os.path.isdir(settings.ZDS_APP['content']['repo_private_path']):
+        #     shutil.rmtree(settings.ZDS_APP['content']['repo_private_path'])
+        # if os.path.isdir(settings.ZDS_APP['content']['repo_public_path']):
+        #     shutil.rmtree(settings.ZDS_APP['content']['repo_public_path'])
+        # if os.path.isdir(settings.MEDIA_ROOT):
+        #     shutil.rmtree(settings.MEDIA_ROOT)
```
Commentaire à enlever ?

AmauryCarrade replied:
À dé-commenter plutôt, mais oui.

Situphen commented on zds/tutorialv2/api/views.py:
```diff
@@ -104,3 +108,33 @@ def post(self, request, *args, **kwargs):
             return Response({}, status=status.HTTP_201_CREATED, headers={})
+
+
+"""
+Lists the most recent exports for this content, and their status.
+"""
+class ExportsView(ListAPIView):
+    serializer_class = PublicationEventSerializer
+    pagination_class=None
```
```suggestion
class ExportsView(ListAPIView):
    """
    Lists the most recent exports for this content, and their status.
    """

    serializer_class = PublicationEventSerializer
    pagination_class=None
```

AmauryCarrade replied:
J'ai corrigé le soucis de PEP8 sur `pagination_class=None` au passage.

Situphen reviewed (commented):
Merci pour ce beau travail !

Quelques commentaires sur la forme suite à une rapide lecture du code.

AmauryCarrade committed:
Commentaires de @Situphen : commentaires superflus

AmauryCarrade commented:
Merci @Situphen, des oublis et inattentions. C'est corrigé !

AmauryCarrade commented:
Cette PR est actuellement bloquée parce que **Travis exécute les tests avec une base MySQL 5.7**, qui ne supporte pas les clauses `OVER`, et ce alors que la prod est en MariaDB 10.4 (ces clauses ont été ajoutées [en version 10.2](https://mariadb.com/kb/en/window-functions-overview/)).

AmauryCarrade committed:
Lint stuff.

- Linted JS & Python files.
- Upgraded tutorialv2/api URLs to `path` at the meantime.

AmauryCarrade committed:
Re-lint as ZdS does not follow black rules.

philippemilink commented:
Rapport de QA
----------------
Je n'ai pas regardé le code, il y a déjà eu une review. Toute la PR fonctionne (sauf l'export PDF, mais c'est normal chez moi).

J'ai quelques remarques:
- pourquoi quand on relance l'export d'un format qui vient d'échouer, ça relance tous les exports ? Je ne dis pas que c'est une mauvaise chose, mais présenté comme tel, c'est surprenant: on s'attend à ce que seulement l'export qu'on vient de relancer soit... relancé. Donc, ce que je ferais: soit le petit lien "Relancer" ne relance effectivement que l'export du format en question, ou alors lors d'un échec, il n'y a pas de petit lien "Relancer" et il faut alors cliquer à nouveau sur le gros bouton.
- le gros bouton pour générer tous les formats ne réapparaît pas une fois les exports terminés. En soit, c'est peut-être une bonne chose (pour éviter que les gens ne cliquent trop dessus), mais peut-être que le bouton pourrait réapparaître lorsque ferme puis rouvre la modale (sans avoir rechargé la page entre temps) ?

Ces remarques se discutent (peut-être pas ici), mais tant que tu as le nez dans ce code... ! 😄 

À part ces remarques, c'est bon pour moi. ✔️

AmauryCarrade commented:
Merci pour la QA !

> pourquoi quand on relance l'export d'un format qui vient d'échouer, ça relance tous les exports ? Je ne dis pas que c'est une mauvaise chose, mais présenté comme tel, c'est surprenant: on s'attend à ce que seulement l'export qu'on vient de relancer soit... relancé. Donc, ce que je ferais: soit le petit lien "Relancer" ne relance effectivement que l'export du format en question, ou alors lors d'un échec, il n'y a pas de petit lien "Relancer" et il faut alors cliquer à nouveau sur le gros bouton.

Effectivement. Ça relance tout essentiellement car on n'a pas d'API qui relance juste un export. J'ai jugé que ce n'était pas trop grave, mais il est possible de retirer le bouton si vous jugez que c'est embêtant. Ou d'ajouter une API juste pour ça mais je trouve que ce n'est pas très utile (et il faudrait alors que je comprenne le système de file d'attente, mais c'est une autre histoire).

> le gros bouton pour générer tous les formats ne réapparaît pas une fois les exports terminés. En soit, c'est peut-être une bonne chose (pour éviter que les gens ne cliquent trop dessus), mais peut-être que le bouton pourrait réapparaître lorsque ferme puis rouvre la modale (sans avoir rechargé la page entre temps) ?

C'était volontaire pour éviter que les gens ne rappuient dessus. En effet, s'iels le font, ça va relancer d'autres demandes dans la file d'attente et les exports vont être générés plusieurs fois, ce pourquoi je masque le bouton une fois les exports demandés.

En réalité le bouton ne disparaît pas complètement : il revient après deux minutes, si on ne recharge pas la page. Ça limite pas mal les usages abusifs tout en permettant de relancer si assez de temps s'est écoulé (bien que ça ne devrait jamais être nécessaire).

philippemilink commented:
> > pourquoi quand on relance l'export d'un format qui vient d'échouer, ça relance tous les exports ? Je ne dis pas que c'est une mauvaise chose, mais présenté comme tel, c'est surprenant: on s'attend à ce que seulement l'export qu'on vient de relancer soit... relancé. Donc, ce que je ferais: soit le petit lien "Relancer" ne relance effectivement que l'export du format en question, ou alors lors d'un échec, il n'y a pas de petit lien "Relancer" et il faut alors cliquer à nouveau sur le gros bouton.
> 
> Effectivement. Ça relance tout essentiellement car on n'a pas d'API qui relance juste un export. J'ai jugé que ce n'était pas trop grave, mais il est possible de retirer le bouton si vous jugez que c'est embêtant. Ou d'ajouter une API juste pour ça mais je trouve que ce n'est pas très utile (et il faudrait alors que je comprenne le système de file d'attente, mais c'est une autre histoire).

Dans ce cas, je serai d'avis, en effet, de retirer le petit lien "Relancer", et ne garder que le gros bouton. 

> > le gros bouton pour générer tous les formats ne réapparaît pas une fois les exports terminés. En soit, c'est peut-être une bonne chose (pour éviter que les gens ne cliquent trop dessus), mais peut-être que le bouton pourrait réapparaître lorsque ferme puis rouvre la modale (sans avoir rechargé la page entre temps) ?
> 
> C'était volontaire pour éviter que les gens ne rappuient dessus. En effet, s'iels le font, ça va relancer d'autres demandes dans la file d'attente et les exports vont être générés plusieurs fois, ce pourquoi je masque le bouton une fois les exports demandés.
> 
> En réalité le bouton ne disparaît pas complètement : il revient après deux minutes, si on ne recharge pas la page. Ça limite pas mal les usages abusifs tout en permettant de relancer si assez de temps s'est écoulé (bien que ça ne devrait jamais être nécessaire).

Ça me va comme argument, on peut garder ça comme ça.

AmauryCarrade committed:
Retrait du bouton “Relancer” en cas d'erreur.

Aussi, reformulation de textes afin d'être plus clair et cohérent avec
l'interface légèrement modifiée.

AmauryCarrade commented:
J'ai retiré le lien et _rebasé_ au passage. QA pour ce petit changement : ouvrir l'interface et vérifier que les éléments échoués n'ont pas le lien, et que les éléments non-échoués l'ont, comme ci-après ; et que le lien des éléments non-échoués pointe bien vers les fichiers (il ne doit pas relancer l'export).

![Interface des exports mise à jour](https://user-images.githubusercontent.com/1417570/86595384-ef5cec00-bf98-11ea-9899-f80cf6b106b9.png)

AmauryCarrade committed:
L'élément ne doit pas être créé du tout en cas d'échec.

philippemilink commented:
Rapport de QA
---------------

Le lien "Relancer" propre à chaque format dont l'export a échoué n'est plus présent et les liens "Télécharger" renvoient correctement vers le contenu exporté.

**QA OK** ✔️
